architecture/architecture.md

```diff
@@ -0,0 +1,115 @@
+2. The compiler converts the generated AST to bytecode.
+3. The VM then executes the bytecode given user supplied input parameters and returns its result.
+
+## Entry points
+
+The entry points are as follows:
+
+- The 'main' method of the application: `run`, located in `src/lib.rs:70`. This method will call the compiler, which in turn will call the parser, and pass the compiled bytecode to the VM.
+- Parser: `parse`, located in `parser/src/parser.rs:74`.
+- Compiler: `compile_top`, located in `compiler/src/compiler.rs:90`.
+- VM: `run_code_obj`, located in `vm/src/rm.rs:459`.
+
+## Modules
+
+Here we give a brief overview of each module and its function. For more details for the separate crates please take a look at their respective READMEs.
+
+### Bytecode
+
+This module (single file at moment) holds the representation of bytecode for RustPython
+<!-- For instance, the following function
+```python
+def f(x):
+    return x + 1
+```
+Is compiled to
+```rust
+// Not sure what this should be yet
+``` -->
+
+### Compiler
+
+Python compilation to bytecode. The interface is exposed through the porcelain crate with `compile` and `compile_symtable`, while the inner workings are defined in compiler/src/compile.rs, which is mostly an adaptation of the CPython implementation.
+
+### Derive
+
+Rust language extensions and macros specific to rustpython. Here we can find the definition of `PyModule` and `PyClass` along with useful macros like `py_compile!`
+
+### Parser
+
+All the functionality required for parsing python sourcecode to an abstract syntax tree (AST)
+1. Lexical Analysis
+2. Parsing
+
+As Python heavily relies on whitespace and indentation to organize code, the crate used for parsing, [LALRPOP](https://github.com/lalrpop/lalrpop), the raw source code is first preprocessed by a lexer which makes sure that `Indent` and `Dedent` tokens occur at the correct locations. Then, the parser recursively generates an AST for the code which can be processed by the compiler.
+
+### Lib
+
+Python side of the standard libary, copied over (with care) from CPython sourcecode.
+
+### Lib/test
+
+
+CPython test suite, which can be used to compare with CPython in terms of functionality and performance.
+Many of these files have been modified to fit with the current state of RustPython (when they were added), with one of three ways:
+
+- The test has been commented out completely if the parser could not create a valid code object (Syntax Error)
+- A test has been marked as `unittest.skip("TODO: RustPython")` if it led to a crash of RustPython
+- A test has been marked as `unittest.expectedFailure` with `TODO: RustPython` left on top if the test can run but failed
+
+Note: This is a recommended route to starting with contributing. To get started please take a look [this blog post](https://rustpython.github.io/guideline/2020/04/04/how-to-contribute-by-cpython-unittest.html).
+
+
+### VM
+
+Python VM
+
+- builtins: all the builtin functions
```

Comment:
Builtin functions live in `make_module.rs` of `vm/src/builtins`. Rest of the files in `builtins` contain the builtin types. Not sure what `obj` is, an older folder that used to exist?